src/internal/m365/graph/middleware.go

```diff
@@ -252,93 +245,88 @@ func (mw RetryMiddleware) retryRequest(
 	priorErr error,
 ) (*http.Response, error) {
-	status := "unknown_resp_status"
-	statusCode := -1
+	ctx = clues.Add(ctx, "retry_count", executionCount)
 
 	if resp != nil {
-		status = resp.Status
-		statusCode = resp.StatusCode
+		ctx = clues.Add(ctx, "prev_resp_status", resp.Status)
 	}
 
-	ctx = clues.Add(
-		ctx,
-		"prev_resp_status", status,
-		"retry_count", executionCount)
-
-	// only retry under certain conditions:
-	// 1, there was an error.  2, the resp and/or status code match retriable conditions.
+	// only retry if all the following conditions are met:
+	// 1, there was a prior error.
+	// 2, the resp and/or status code match retriable conditions.
 	// 3, the request is retriable.
 	// 4, we haven't hit our max retries already.
-	if (priorErr != nil || mw.isRetriableRespCode(ctx, resp, statusCode)) &&
+	shouldRetry := (priorErr != nil || mw.isRetriableRespCode(ctx, resp)) &&
 		mw.isRetriableRequest(req) &&
-		executionCount < mw.MaxRetries {
-		executionCount++
+		executionCount < mw.MaxRetries
 
-		delay := mw.getRetryDelay(req, resp, exponentialBackoff)
-		cumulativeDelay += delay
+	if !shouldRetry {
+		return resp, stackReq(ctx, req, resp, priorErr).OrNil()
+	}
 
-		req.Header.Set(retryAttemptHeader, strconv.Itoa(executionCount))
+	executionCount++
 
-		timer := time.NewTimer(delay)
+	delay := mw.getRetryDelay(req, resp, exponentialBackoff)
+	cumulativeDelay += delay
 
-		select {
-		case <-ctx.Done():
-			// Don't retry if the context is marked as done, it will just error out
-			// when we attempt to send the retry anyway.
-			return resp, clues.Stack(ctx.Err()).WithClues(ctx)
+	req.Header.Set(retryAttemptHeader, strconv.Itoa(executionCount))
 
-		case <-timer.C:
-		}
+	timer := time.NewTimer(delay)
 
-		// we have to reset the original body reader for each retry, or else the graph
-		// compressor will produce a 0 length body following an error response such
-		// as a 500.
-		if req.Body != nil {
-			if s, ok := req.Body.(io.Seeker); ok {
-				_, err := s.Seek(0, io.SeekStart)
-				if err != nil {
-					return nil, Wrap(ctx, err, "resetting request body reader")
-				}
-			}
-		}
+	select {
+	case <-ctx.Done():
+		// Don't retry if the context is marked as done, it will just error out
+		// when we attempt to send the retry anyway.
+		return resp, clues.Stack(ctx.Err()).WithClues(ctx)
 
-		nextResp, err := pipeline.Next(req, middlewareIndex)
-		if err != nil && !IsErrTimeout(err) && !IsErrConnectionReset(err) {
-			return nextResp, stackReq(ctx, req, nextResp, err)
-		}
+	case <-timer.C:
+	}
 
-		return mw.retryRequest(
-			ctx,
-			pipeline,
-			middlewareIndex,
-			req,
-			nextResp,
-			executionCount,
-			cumulativeDelay,
-			exponentialBackoff,
-			err)
+	// we have to reset the original body reader for each retry, or else the graph
+	// compressor will produce a 0 length body following an error response such
+	// as a 500.
+	if req.Body != nil {
+		if s, ok := req.Body.(io.Seeker); ok {
```

Comment:
I know this isn't code directly modified in this PR, but should we be logging an error or something if we don't have an `io.Seeker`?

Reply:
I don't see any harm in it.